crates/holochain_p2p/src/spawn/actor.rs

```diff
@@ -726,76 +880,184 @@ impl kitsune_p2p::event::KitsuneP2pEventHandler for HolochainP2pActor {
+                    $($rest)*
+                );
+            }
+            Err(err) => {
+                tracing::trace!(
+                    target: "NETAUDIT",
+                    m = "holochain_p2p",
+                    ?err,
+                    elapsed_s = __elapsed_s,
+                    $($rest)*
+                );
             }
-            __out
         }
-        .boxed()
-        .into())
+        __out
     }};
 }
------------------------ */
 
 impl actor::HcP2p for HolochainP2pActor {
     fn join(
         &self,
-        _dna_hash: DnaHash,
-        _agent_pub_key: AgentPubKey,
+        dna_hash: DnaHash,
+        agent_pub_key: AgentPubKey,
         _maybe_agent_info: Option<AgentInfoSigned>,
     ) -> BoxFut<'_, HolochainP2pResult<()>> {
-        Box::pin(async move { todo!() })
+        Box::pin(async move {
+            let kitsune = self.kitsune()?;
+            let space = kitsune.space(dna_hash.to_k2_space()).await?;
+
+            let local_agent: DynLocalAgent = Arc::new(HolochainP2pLocalAgent::new(
+                agent_pub_key,
+                DhtArc::FULL,
+                self.lair_client.clone(),
+            ));
+
+            space.local_agent_join(local_agent).await?;
+            Ok(())
+        })
     }
 
     fn leave(
         &self,
-        _dna_hash: DnaHash,
-        _agent_pub_key: AgentPubKey,
+        dna_hash: DnaHash,
+        agent_pub_key: AgentPubKey,
     ) -> BoxFut<'_, HolochainP2pResult<()>> {
-        Box::pin(async move { todo!() })
+        Box::pin(async move {
+            let kitsune = self.kitsune()?;
+            let space = kitsune.space(dna_hash.to_k2_space()).await?;
+
+            space.local_agent_leave(agent_pub_key.to_k2_agent()).await;
+            Ok(())
+        })
     }
 
     fn call_remote(
         &self,
-        _dna_hash: DnaHash,
-        _to_agent: AgentPubKey,
-        _zome_call_params_serialized: ExternIO,
-        _signature: Signature,
+        dna_hash: DnaHash,
+        to_agent: AgentPubKey,
+        zome_call_params_serialized: ExternIO,
+        signature: Signature,
     ) -> BoxFut<'_, HolochainP2pResult<SerializedBytes>> {
-        Box::pin(async move { todo!() })
+        Box::pin(async move {
+            let kitsune = self.kitsune()?;
+            let space_id = dna_hash.to_k2_space();
+            let space = kitsune.space(space_id.clone()).await?;
+
+            let byte_count = zome_call_params_serialized.0.len();
+
+            let to_url = space
+                .peer_store()
+                .get(to_agent.to_k2_agent())
+                .await?
+                .and_then(|i| i.url.clone())
+                .ok_or_else(|| HolochainP2pError::other("call_remote: no url for peer"))?;
+
+            let (msg_id, req) = crate::wire::WireMessage::call_remote_req(
+                to_agent,
+                zome_call_params_serialized,
+                signature,
+            );
+            // someday we may support batching
```

Comment:
A little tempted to not have code that supports this for now. Most of these should be in response to a timer or a user action and won't be coming in as a batch. I think if we want to get more data from peers at once that should happen with pub/sub and we'd probably have to extend this code to support that anyway. What do you think?

Reply:
The messagepack overhead for wrapping a single message (or any count < 16) in a vec is a single byte, and making the decode logic be a loop instead of a single operation doesn't feel like it adds much complexity.

The original RemoteSignal code seemed to be trying to set it up such that if multiple messages came in at once destined to the same peer node that they could be batched (though it wasn't doing that in reality). I've deleted that entirely for now, but it feels like there could be strategic operations like that which could greatly benefit once we get to the stage of optimizing.

I'm also fine if we want to say no for now, and that will require a protocol break. Let me know if we should talk about this is realtime.

Reply:
I'm torn. I want to keep this as simple as possible, but I agree we should work towards minimizing protocol breaks.

It's here, it's low overhead as you say. Let's go with it for now :)